modelling/src/test/java/org/axonframework/modelling/saga/repository/CachingSagaStoreTest.java

```diff
@@ -219,4 +224,43 @@ StubSaga.class, sagaId, mock(StubSaga.class), associationValues
             fail("An unexpected exception occurred during concurrent invocations on the CachingSagaStore.", e);
         }
     }
+
+    @Test
+    void canHandleConcurrentReadsAndWritesThroughAnnotatedSagaRepository() {
+        SagaRepository<StubSaga> sagaRepository = AnnotatedSagaRepository.<StubSaga>builder()
+                                                                         .sagaType(StubSaga.class)
+                                                                         .sagaStore(testSubject)
+                                                                         .build();
+        int concurrentOperations = 64;
+        ExecutorService executor = Executors.newFixedThreadPool(16);
+        AssociationValue associationValue = new AssociationValue("StubSaga-id", "value");
+
+        try {
+            IntStream.range(0, concurrentOperations)
+                     .mapToObj(i -> CompletableFuture.runAsync(
+                             () -> {
+                                 try {
+                                     UnitOfWork<Message<?>> uow = DefaultUnitOfWork.startAndGet(null);
+                                     String sagaId = IdentifierFactory.getInstance().generateIdentifier();
+                                     // Create instances
+                                     Saga<StubSaga> saga = sagaRepository.createInstance(sagaId, StubSaga::new);
+                                     uow.execute(() -> saga.getAssociationValues().add(associationValue));
+                                     // Find Saga identifiers
+                                     Set<String> sagaIds = sagaRepository.find(associationValue);
+                                     // Load Sagas
+                                     DefaultUnitOfWork.startAndGet(null)
+                                                      .execute(() -> sagaIds.forEach(sagaRepository::load));
+                                 } catch (Exception e) {
+                                     throw new RuntimeException(e);
+                                 }
+                             },
+                             executor
+                     ))
+                     .reduce(CompletableFuture::allOf)
+                     .orElse(CompletableFuture.completedFuture(null))
+                     .get(30, TimeUnit.SECONDS);
```

Comment:
64 operations and a timeout of 30 seconds seem a bit long to me, could we make it smaller and faster?

Reply:
Sounds fair to me. Changed the concurrency count to 32 and timeout to 10